design/ceph-openshift-fencing-mitigation.md

```diff
@@ -0,0 +1,52 @@
+# Handling Openshift Fencing through managed MachineDisruptionBudgets
+
+## Goals
+
+- Use the MachineDisruptionBudget to ensure that OCP 4.x style fencing does not cause data unavailability and loss.
+
+## What is fencing in OCP 4.x?
+
+Openshift uses `Machines` and `MachineSets` from the [cluster-api](https://github.com/kubernetes-sigs/cluster-api) to dynamically provisions nodes. Fencing is a remediation method that reboots/deletes `Machine` CRDs to solve problems with automatically provisioned nodes.
+
+Once [MachineHealthCheck controller](https://github.com/openshift/machine-api-operator#machine-healthcheck-controller) detects that a node is `NotReady` (or some other configured condition), it will remove the associated `Machine` which will cause the node to be deleted. The `MachineSet` controller will then replace the `Machine` via the machine-api. The exception is on baremetal platforms where fencing will reboot the underlying `BareMetalHost` object instead of deleting the `Machine`.
+
+
+## Why can't we use `PodDisruptionBudget`?
+
+Fencing does not use the eviction api. It is for `Machine`s and not `Pod`s.
+
+## Will we need to do large storage rebalancing after fencing?
+
+Hopefully not. On cloud platforms, the OSDs can be rescheduled on new nodes along with their backing PVs, and on baremetal where the local PVs are tied to a node, fencing will simply reboot the node instead of destroying it.
+
+
+# Problem Statement
+We need to ensure that only one node can be fenced at a time and that Ceph is fully recovered (has PGs clean) before any fencing is initiated. The available pattern for limiting fencing is the MachineDisruptionBudget which allows us to specify maxUnavailable. However, this won’t be sufficient to ensure that Ceph has recovered before fencing is initiated.
+
```

Comment:
Why exactly won't this be sufficient?

Reply:
Is this because the disruption budget doesn't understand waiting for ceph health status?

Reply:
I'll clarify in the doc